In [1]:
import pandas as pd 

In [2]:
df =  pd.read_csv('For_Machine_Learning')

In [5]:
df = df.drop(columns={'Unnamed: 0'})

In [6]:
df.head()

Income  Age  Experience  Married/Single  House_Ownership  Car_Ownership  \
0  1303834   23           3               0                0              0   
1  7574516   40          10               0                0              0   
2  3991815   66           4               1                0              0   
3  6256451   41           2               0                0              1   
4  5768871   47          11               0                0              0   

   Profession  CITY  STATE  CURRENT_JOB_YRS  CURRENT_HOUSE_YRS  Risk_Flag  
0          33   250     13                3                 13          0  
1          43   226     14                9                 13          0  
2          47     8     12                4                 10          0  
3          43    53     17                2                 12          1  
4          11   295     22                3                 14          1

In [7]:
# seprating into x and y 

In [8]:
x= df.drop(columns='Risk_Flag' , axis=1)

In [9]:
y = df['Risk_Flag']

In [10]:
y.value_counts()

0    221004
1     30996
Name: Risk_Flag, dtype: int64

In [11]:
# now using smote function of over_sampling 

In [12]:
from imblearn.over_sampling import SMOTE

In [13]:
smote  =  SMOTE()

In [14]:
x, y = smote.fit_resample(x, y)

In [15]:
y.value_counts()

0    221004
1    221004
Name: Risk_Flag, dtype: int64

In [16]:
# so we totally over sample the data 

In [17]:
# now splitting the dtaa set into training and testing form 

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
x_train , x_test , y_train , y_test = train_test_split(x,y , test_size=0.2 , random_state=42)

In [20]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(353606, 11)
(88402, 11)
(353606,)
(88402,)


# Finding the best model 

In [21]:
from sklearn.linear_model import LogisticRegression

In [22]:
lo =  LogisticRegression()
lo.fit(x_train , y_train)
y_pred1 = lo.predict(x_test)

In [23]:
from sklearn.metrics import accuracy_score

In [24]:
score1 =  accuracy_score(y_test , y_pred1)

In [25]:
score1

0.501380059274677

In [26]:
from sklearn.tree import DecisionTreeClassifier

In [27]:
dt =  DecisionTreeClassifier()
dt.fit(x_train , y_train)
y_pred2 =  dt.predict(x_test)

In [28]:
score2 =  accuracy_score(y_test , y_pred2)

In [29]:
score2

0.908124250582566

In [30]:
from sklearn.ensemble import RandomForestClassifier

In [31]:
rf = RandomForestClassifier()
rf.fit(x_train , y_train)


RandomForestClassifier()

In [33]:
y_pred3 =  rf.predict(x_test)

In [34]:
score3 = accuracy_score(y_test , y_pred3)

In [35]:
score3

0.9257256623153323

In [37]:
# Accuracy scrore without dropping city columns 
print("Score of Logistic Regression are " , score2)
print("Score of DecisionTreeClassifier are " , score2)
print("Score of Random forest Classifier  are " , score3)


Score of Logistic Regression are  0.908124250582566
Score of DecisionTreeClassifier are  0.908124250582566
Score of Random forest Classifier  are  0.9257256623153323


# conclusion

After applying the oversampling on dataset and without dropping any columns our accuracy score of random forest classifier
are high as compare to another model that is 92.57% 

In [40]:
# testing our model
df.columns

Index(['Income', 'Age', 'Experience', 'Married/Single', 'House_Ownership',
       'Car_Ownership', 'Profession', 'CITY', 'STATE', 'CURRENT_JOB_YRS',
       'CURRENT_HOUSE_YRS', 'Risk_Flag'],
      dtype='object')

In [41]:
df.head(1)

Income  Age  Experience  Married/Single  House_Ownership  Car_Ownership  \
0  1303834   23           3               0                0              0   

   Profession  CITY  STATE  CURRENT_JOB_YRS  CURRENT_HOUSE_YRS  Risk_Flag  
0          33   250     13                3                 13          0

In [42]:
test_data = pd.DataFrame({'Income':130383,
                          'Age':23,
                          'Experience':3,
                          'Married/Single':0,
                           'House_Ownership':0, 
                          'Car_Ownership':0,
                          'Profession':33, 
                          'CITY':250, 
                          'STATE':13,
                          'CURRENT_JOB_YRS':3,
                          'CURRENT_HOUSE_YRS':13,
                         
                        } , index=[0])

In [43]:
test_data

Income  Age  Experience  Married/Single  House_Ownership  Car_Ownership  \
0  130383   23           3               0                0              0   

   Profession  CITY  STATE  CURRENT_JOB_YRS  CURRENT_HOUSE_YRS  
0          33   250     13                3                 13

In [44]:
p=rf.predict(test_data)
if p==0:
    print("repay the loan ")
else:
    print("Fail to repay the loan")

repay the loan 


In [45]:
# hence our model was working totally fine 

In [46]:
# Saving the model 

In [47]:
import joblib

In [48]:
joblib.dump(rf , 'model_over_sampling_done')

['model_over_sampling_done']

In [49]:
model = joblib.load('model_over_sampling_done')

In [50]:
model.predict(test_data)

array([0], dtype=int64)

In [51]:
# hence solve problem 

In [52]:
#for future prediction model_over_sampling_done is work perfectly fine